### ISSUE TITLE
storage: judge transactions interact before multiraft replication

### ISSUE THREAD

**Issue by wuweiweiky** [cockroachdb/cockroach#2632]
At present, the transactions conflicts  are  judged in "mvccPutInternal" method when  applying multiraft commited entry.  
I wonder whether the transactions conflicts can be judged before multiraft replication, if we do so, unnecessary multiraft replication can be avoided.

**Comment by tschottdorf** [comment]
there are some optimizations we could go for (for example #627, implementation of which has been shelved because such optimizations are likely to do more harm than good at this stage) but at the end, I think at least once we must wait for a conflict to arise, which means that you need to check something consistently, which means going through Raft. Did you have anything specific in mind?

**Comment by wuweiweiky** [comment]
I think we can judge transaction consistency before multiraft replication. If there is a conflict, restart or abort the current transaction. If there no conflict, just send to multiraft and wait for applying. 

At this pattern, there  is  a  special  scene we need  to pay attention to, that is when raft leader changed, we must wait committed entry of new leader  applied in order to handle transaction conflict. 

Maybe there are some other reasons prevent this optimization .

**Comment by tschottdorf** [comment]
you're suggesting just reading straight from the engine without going through Raft in the hopes of taking a shortcut in conflict resolution? That may or may not be worth it and it's similar in spirit to #627:
If you check a value you're about to write and there's an apparent conflict (for example, an intent) you can only be sure it's going to remain a conflict if it's not an intent (after all, the intent could be aborted before you get to write). If such conflicts are relatively rare (and they should be) you're going to effectively double the number of calls down to the engine - not worth it.
Even just using the timestamp of a serializable operation is not trivial: you'd think that we should be able to return a `TransactionRetryError` immediately when the timestamp cache has something newer, without hitting the engine at all (maybe that's what you were suggesting?) - but even there's a caveat: The timestamp cache could have been increased due to writing an intent, which our transaction needs to be able to run into in order to Push. So for that optimization (which might be worthwhile) you'd have to look into avoiding updating the timestamp cache when an intent is written (updating only when it's resolved).

The bottom line is, there's certainly **something** to win there, but we're still mostly looking at correctness, and will put more complicated things in later. I'd however like to encourage you to consider these questions and share your insights with us - we'll get to a stage at which we're optimizing soon.

**Comment by tschottdorf** [comment]
Closing for #627.
